dvc/repo/reproduce.py

```diff
@@ -76,25 +74,15 @@ def _track_stage(stage: "Stage") -> None:
     return context.track_changed_files()
 
 
-def _track_top_level(repo: "Repo") -> None:
-    context = repo.scm_context
-    for metric_file in _collect_top_level_metrics(repo):
-        context.track_file(metric_file)
-    for param_file in _collect_top_level_params(repo):
-        context.track_file(param_file)
-    for plot_file in repo.index._plot_sources:  # pylint: disable=protected-access
-        context.track_file(plot_file)
-    context.track_changed_files()
-
-
 @locked
 @scm_context
-def reproduce(  # noqa: C901
+def reproduce(  # noqa: C901, PLR0912
     self: "Repo",
     targets=None,
     recursive=False,
     pipeline=False,
     all_pipelines=False,
+    after_repro_callback=None,
```

Comment:
`reproduce` is running in a `scm_context`, so I had two choices here: Either extract the logic out and use callbacks, or run this conditionally through some keyword arguments.

I went with the former, as I believe we should have more callbacks in `reproduce`, and fit nicely than adding more keyword arguments.

Reply:
Users will miss `scm_context` logging if we just added them outside `reproduce` or duplicated outputs if we use `scm_context` again outside.

Reply:
@skshetry what are the main benefits of the current `scm_context` usage as a decorator?

Reply:
The main use is to show message like following:
```
To track the changes with git, run:

	git add foo.dvc .gitignore
```

for all the modifications that dvc makes to the user's workspace. 

The reason why it is a decorator is because we treat each `Repo` API as a separate session, so they are equivalent to one CLI invocation. We want to show that message for individual CLI sessions.

`exp run` breaks this by relying on `reproduce`.

Reply:
Also in a way, we are misusing `scm_context` in the context of `exp run`. We want to stage them so that experiments can track them behind the scene, but also show this message for workspace (or autostage if it set).

Reply:
@skshetry just to clarify, do you think `exp apply` should stage the experiment changes when `core.autostage` is set?

Reply:
I'd say no, as `apply` is a passive operation and is intended to just restore to workspace. And what `exp apply` behavior is expected overrides config's.

If we autostage, what would be the way to avoid that? I feel like `--staged/--cached` might be better/explicit for `exp apply` than respecting `core.autostage`. 

Anyway, this is just my opinion. cc'ing @dberenbaum for his opinion here.

Reply:
Thanks @skshetry. No strong opinion here. Let's see if it comes up from users.